src/include/platform/internal/GenericConfigurationManagerImpl.ipp

```diff
@@ -249,7 +250,39 @@ CHIP_ERROR GenericConfigurationManagerImpl<ConfigClass>::Init()
     ReturnErrorOnFailure(err);
 
     err = StoreUniqueId(uniqueId, strlen(uniqueId));
+    ReturnErrorOnFailure(err);
+
+    bool failSafeArmed;
+
+    // If the fail-safe was armed when the device last shutdown, initiate cleanup based on the pending Fail Safe Context with
+    // which the fail-safe timer was armed.
+    if (GetFailSafeArmed(failSafeArmed) == CHIP_NO_ERROR && failSafeArmed)
+    {
+        FabricIndex fabricIndex;
+        bool addNocCommandInvoked;
+        bool updateNocCommandInvoked;
+
+        ChipLogProgress(DeviceLayer, "Detected fail-safe armed on reboot");
+
+        err = FailSafeContext::LoadFromStorage(fabricIndex, addNocCommandInvoked, updateNocCommandInvoked);
+        SuccessOrExit(err);
+
+        ChipDeviceEvent event;
+        event.Type                                                = DeviceEventType::kFailSafeTimerExpired;
+        event.FailSafeTimerExpired.PeerFabricIndex                = fabricIndex;
+        event.FailSafeTimerExpired.AddNocCommandHasBeenInvoked    = addNocCommandInvoked;
+        event.FailSafeTimerExpired.UpdateNocCommandHasBeenInvoked = updateNocCommandInvoked;
+
+        err = PlatformMgr().PostEvent(&event);
+        SuccessOrExit(err);
+
+        DeviceControlServer::DeviceControlSvr().GetFailSafeContext().SetFailSafeBusy(true);
```

Comment:
As a followup: We should still have the event posting and `SetFailSafeBusy` all be inside the fail-safe, so we don't duplicate this logic.

Reply:
https://github.com/project-chip/connectedhomeip/issues/16814